lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -815,6 +889,9 @@ class PersistState s where
     selectState :: (W.WalletId, W.SlotId) -> SqlPersistT IO (Maybe s)
     -- | Remove the state for all checkpoints of a wallet.
     deleteState :: W.WalletId -> SqlPersistT IO ()
+    -- | Remove the state for one checkpoint of a wallet.
+    deleteCheckpointState :: W.WalletId -> W.SlotId -> SqlPersistT IO ()
```

Comment:
There seems to be some duplication between `deleteCheckpoints` and `deleteCheckpoint` which ripples down to here and `deleteState`.  We only use `deleteCheckpoints` and `deleteState` when removing an entire wallet.

What about simply allowing an extra parameter to deleteState and deleteCheckpoints to support passing one or several slot ids:

```hs
-- | Delete the checkpoint associated with a wallet and slotId.
deleteCheckpoints
    :: forall s. PersistState s
    => W.WalletId
    -> [W.SlotId]
    -> SqlPersistT IO ()
deleteCheckpoints wid sids = do
    deleteWhere [UtxoWalletId ==. wid, UtxoSlot <-. sids]
    deleteWhere [CheckpointWalletId ==. wid, CheckpointSlot <-. sids]
    deleteState @s wid sids

deleteState :: W.WalletId -> [W.SlotId] -> SqlPersistT IO ()
deleteState wid sids = do
    cpid <- selectList
        [ SeqStateCheckpointWalletId ==. wid
        , SeqStateCheckpointSlot <-. sids
        ] []
    deleteWhere
        [ SeqStatePendingIxCheckpointId <-. fmap entityKey cpid
        ]
    deleteCascadeWhere
        [ SeqStateCheckpointWalletId ==. wid
        , SeqStateCheckpointSlot <-. sids
        ]
    deleteWhere
        [ SeqStateAddressWalletId ==. wid
        , SeqStateAddressSlot <-. sids
        ]
```

Then, when deleting a wallet, we simply need to first query all the slot ids of all checkpoints, and make sure to pass them as parameter. It's not the most efficient, but for removing wallet which is a fairly rare operation, it sounds pretty okay IMO.

Reply:
I changed delete checkpoints to take Maybe SlotId because It's either one checkpoint or all that we want to delete.